### PR TITLE
arm64: dts: qcom: Add dt for Samsung Grand Prime

### PR DESCRIPTION
**Pull request by SiddharthManthan**
This pull request adds support for gprime and fortuna3g devices.

Add initial support for :
* gpio keys and WCNSS
* ARMv7 device tree for gprime
* Vibrator
* rt5033 fuelgauge. Also compatible with sec-fuelgauge
* Zinitix BT541 with touchkey
* Modem and sound
* Audio jack

The following pull requests need to merged first :
- [ ] Commits from for-pwm and for-clk branches of https://github.com/TravMurav/linux
- [ ] https://github.com/msm8916-mainline/lk2nd/pull/151
- [ ] https://github.com/msm8916-mainline/linux-panel-drivers/pull/10

ToDo :
- [ ] Enable clk-pwm in msm8916_defconf

### CODE REVIEW

File: arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts [hunk 1]
```diff
@@ -8,4 +8,63 @@
...
+		clocks = <&gcc GCC_GP2_CLK>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&backlight_pwm_default>;
+	};
+};
+
+&dsi0 {
+	panel@0 {
+		compatible = "samsung,s6d78a0-gh9607501a";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts:36]: The driver for `samsung,s6d78a0-gh9607501a` is not found in this PR?

**Review by wonderfulShrineMaidenOfParadise — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts (hunk 1 above)
Comment (SiddharthManthan) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts:36]: I added it at https://github.com/msm8916-mainline/linux-panel-drivers/pull/10. should be available after next kernel update.

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi [hunk 2]
```diff
@@ -107,6 +107,30 @@
 	};
 };
 
+&blsp_i2c2 {
+	status = "okay";
+
+	accelerometer@1d {
+		compatible = "st,lis2hh12";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi:114]: > arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add accelerometer

Maybe reword this :p
Some variants e.g. SM-G530FZ/W use `bosch,bmc150_accel` instead of this.

**Review by wonderfulShrineMaidenOfParadise — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts [hunk 3]
```diff
@@ -21,6 +21,32 @@
...
+			interrupt-parent = <&msmgpio>;
+			interrupts = <21 IRQ_TYPE_EDGE_RISING>;
+
+			enable-gpios = <&msmgpio 20 GPIO_ACTIVE_HIGH>;
+			firmware-gpios = <&msmgpio 49 GPIO_ACTIVE_HIGH>;
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&nfc_default>;
+		};
+	};
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts:49]: NFC varies a lot:
- SM-G530FZ: s3fwrn5 on I2C GPIO pin 0 & 1
- SM-G530H: No NFC
- SM-G530W: s3fwrn5 on I2C6
- SM-G530Y: pn547 on I2C GPIO pin 0 & 1

**Review by wonderfulShrineMaidenOfParadise — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts [hunk 4]
```diff
@@ -21,4 +21,25 @@
...
+				remote-endpoint = <&dsi0_out>;
+			};
+		};
+	};
+};
+
+&dsi0_out {
+	data-lanes = <0 1>;
+	remote-endpoint = <&panel_in>;
+};
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts:79]: This is meant to be common, so that we can select panel in lk2nd since there are 3 variants:
- HX8389C
- HX8389C_GH9607501A
- S6D78A0_GH9607501A

https://github.com/msm8916-mainline/lk2nd/pull/143/commits/41a528b9200e56a109f9ef6f611008c442ed199b#diff-2f45d9b494a1d6ebd655f6a04cb57e77a8557037104a60c4ab020feafa3dba03R23-R29

Otherwise we don't need this. I cannot actually make decision on this, though.

**Comment by wonderfulShrineMaidenOfParadise** [comment]
@buddyjojo please consider join us :D
Most of stuffs are done. Just need to work on panel, accelerometer and NFC for SM-G530W.

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts [hunk 5]
```diff
@@ -2,179 +2,23 @@
 
 /dts-v1/;
 
-#include "msm8916-pm8916.dtsi"
-#include <dt-bindings/gpio/gpio.h>
-#include <dt-bindings/interrupt-controller/irq.h>
+#include "msm8916-samsung-gprime-common.dtsi"
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts:5]: Not sure whether we need this. The original branch is already generic for all variants.

Comment (SiddharthManthan) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts:5]: @TravMurav

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi (hunk 2 above)
Comment (SiddharthManthan) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi:114]: Reworded to "Add support for ST LIS2HH12 accelerometer."

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts (hunk 3 above)
Comment (SiddharthManthan) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts]: Yes. That's why I added it to the device tree specific to your device.

Comment (wonderfulShrineMaidenOfParadise) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-gprime.dts]: My device has it's own quirk to solve so I have opened #222 instead of working here directly. By the way it will be renamed to `samsung-fortunaltezt` with dedicated device tree.

File: arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi (hunk 2 above)
Comment (wonderfulShrineMaidenOfParadise) [reply arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi:114]: Well I mean the title needed to be changed since it was `samsung-fortunaltezt`. New commit message looks right as well, though.
Still the accelerometer cannot be a common part. I guess that it needs to be added in each device tree.
